compiler/rustc_trait_selection/src/traits/wf.rs

```diff
@@ -704,8 +705,51 @@ impl<'a, 'tcx> TypeVisitor<TyCtxt<'tcx>> for WfPredicates<'a, 'tcx> {
                 ));
             }
 
-            ty::Pat(subty, _) => {
+            ty::Pat(subty, pat) => {
                 self.require_sized(subty, ObligationCauseCode::Misc);
+                match *pat {
+                    ty::PatternKind::Range { start, end, include_end: _ } => {
+                        let mut check = |c| {
+                            let cause = self.cause(ObligationCauseCode::Misc);
+                            self.out.push(traits::Obligation::with_depth(
+                                tcx,
+                                cause.clone(),
+                                self.recursion_depth,
+                                self.param_env,
+                                ty::Binder::dummy(ty::PredicateKind::Clause(
+                                    ty::ClauseKind::ConstArgHasType(c, subty),
+                                )),
+                            ));
+                            if !tcx.features().generic_pattern_types() {
+                                let c = self.tcx().normalize_erasing_regions(
+                                    self.infcx.typing_env(self.param_env),
+                                    c,
+                                );
+                                if c.has_aliases() || c.has_param() {
```

Comment:
This is a bit fishy, but I needed more than `ClauseKind::ConstEvaluatable`. `ConstEvaluatable` only checks that the constant itself will not fail to evaluate, which a generic parameter will also not (due to how we're enforcing things with const generics elsewhere).

So I took the knowledge that these constants always have concrete types and will normalize to error constants if they have errors themselves, to just evaluate here and check if there are still params or aliases around after normalization.

Oh and I need to normalize, can't just check for aliases and params, since we treat any unevaluated const as an alias, even if it refers to a free const. Which I guess makes sense as we want to cause normalization to actually get to the const and evaluate it.

Reply:
I think you can/should just use `c.has_param()` and forget about even normalizing- min const generics doesn't permit `Foo<{ <T as Trait>::USIZE }>` where the assoc const can be normalized to `usize` regardless of `T` so I think design wise it's fine to not be "clever" here and just say that referring to generic parameters syntactically is forbidden.

Reply:
I'm not sure why you need to check for aliases :thinking:

Reply:
uh yea, i think I just thoroughly confused myself. Changed it to avoid normalization and just checking for params. No params in the args of the unevaluated const is what we actually are interested in after all